lib/fibonacci.rb

```diff
@@ -1,8 +1,18 @@
 # Improved Fibonacci
 
-# Time Complexity - ?
-# Space Complexity - ?  (should be O(n))
+# Time Complexity - O(n)
+# Space Complexity - O(n)
 # Hint, you may want a recursive helper method
 def fibonacci(n)
- 
+  if n in results
+    return results(n)
+  else
+    if n == 0
+      return 0
+      if n == 1 
+        return 1
+      end
+    end
+  end
+  results[n] = fib(n-1) + fib(n-2)
```

Comment:
Ok some serious issues here:

1.  There is no method called `fib`
2.  There is no object called `results`
3.  It looks like this is the bad 2<sup>n</sup> solution.

More to work on here.  Take a look at the examples in the notes.